### PR TITLE
make NewtonsoftJson and System.Text.Json works side by side.

### PR DESCRIPTION
**Pull request by liuliang-wt**
The idea is instead of using AddNewtonsoftJson to all controllers,make an attribute to use Newtensoft as Object result's formatter, and apply this attribute to all controllers that in Elsa.Server.Api namespace.

### CODE REVIEW

**Comment by liuliang-wt** [comment]
@dotnet-policy-service agree

**Comment by sfmskywalker** [comment]
@liuliang-wt This is awesome, I didn't know you could do it like this. Very cool!

File: src/server/Elsa.Server.Api/Attributes/NewtonsoftJsonFormatterAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+using System.Buffers;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Elsa.Server.Api.Attributes
+{
+    /// <summary>
+    /// mark a controller to use Newtonsoft as the formatter
```
Comment (sfmskywalker) [inline src/server/Elsa.Server.Api/Attributes/NewtonsoftJsonFormatterAttribute.cs:23]: Please mind capitalization ("Mark a controller ...").

File: src/server/Elsa.Server.Api/Attributes/NewtonsoftJsonFormatterAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            {
+                var jsonOptions = context.HttpContext.RequestServices.GetService<IEndpointContentSerializerSettingsProvider>();
+
+                objectResult.Formatters.RemoveType<SystemTextJsonOutputFormatter>();
+                objectResult.Formatters.Add(new NewtonsoftJsonOutputFormatter(
+                    jsonOptions?.GetSettings(),
+                    context.HttpContext.RequestServices.GetRequiredService<ArrayPool<char>>(),
+                    context.HttpContext.RequestServices.GetRequiredService<IOptions<MvcOptions>>().Value));
+            }
+            //for JsonResult, there is no way to change the formatter, so change them to Object result
```
Comment (sfmskywalker) [inline src/server/Elsa.Server.Api/Attributes/NewtonsoftJsonFormatterAttribute.cs:57]: Please mind capitalization ("For JsonResult...").

File: src/server/Elsa.Server.Api/Attributes/NewtonsoftJsonFormatterAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            else if (context.Result is JsonResult jr)
+            {
+                ObjectResult obj = new ObjectResult(jr.Value);
```
Comment (sfmskywalker) [inline src/server/Elsa.Server.Api/Attributes/NewtonsoftJsonFormatterAttribute.cs:60]: Please use `var`.

File: src/server/Elsa.Server.Api/Extensions/ServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -30,7 +31,8 @@ public static IServiceCollection AddElsaApiEndpoints(this IServiceCollection ser
 
             var setupNewtonsoftJson = apiOptions.SetupNewtonsoftJson ?? (_ => { });
 
-            services.AddControllers().AddNewtonsoftJson(setupNewtonsoftJson);
+            //don't set Newtonsoft globally
```
Comment (sfmskywalker) [inline src/server/Elsa.Server.Api/Extensions/ServiceCollectionExtensions.cs:34]: Please mind capitalization ("// Don't set...").

File: src/server/Elsa.Server.Api/Extensions/ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -54,7 +56,11 @@ public static IServiceCollection AddElsaApiEndpoints(this IServiceCollection ser
                 .AddSingleton<IEndpointContentSerializerSettingsProvider, EndpointContentSerializerSettingsProvider>()
                 .AddAutoMapperProfile<AutoMapperProfile>()
                 .AddSignalR();
-                
+            services.AddMvc(options =>
+            {
+                //use this convertion to set ElsaNewtonsoftJsonConvention to all controllers in Elsa.Server.Api
```
Comment (sfmskywalker) [inline src/server/Elsa.Server.Api/Extensions/ServiceCollectionExtensions.cs:61]: Please mind capitalization ("// Use this...").

There's a small typo: "convertion" => "convention").

**Comment by liuliang-wt** [comment]
I modified the logic abit, and tested it on my own project which uses System.Text.Json, it works :)

**Review by sfmskywalker — APPROVED** [review]

**Comment by liuliang-wt** [comment]
@sfmskywalker Build failed, I checked the log and found NewtonsoftJsonOutputFormatter's constructor is different in older version of Newtensoft, should I change it and submit another PR?

**Comment by sfmskywalker** [comment]
Let’s make sure it uses the latest version of Newtonsoft and update the code as necessary. A new PR would be great 👍🏻

**Comment by liuliang-wt** [comment]
I re-submitted a pr #4441 , it should be ok now

**Pull request merged**